File: For Fun/Assignment 3/Swift version/Exercise 8/Deck.swift
//
//  Deck.swift
//  Temp
//
//  Created by Sunny Buttar on 2016-12-16.
//  Copyright © 2016 Sarpreet Singh. All rights reserved.
//

import Foundation

public class Deck {
    
    private var deck = [Card]()
    
    init() {
        for suit in Card.Suit.allValues {
            for rank in Card.Rank.allValues {
                deck.append(Card(suit: suit, rank: rank))
            }
        }
    }
    
    public func shuffle() {
        if deck.count != 52 {
            print("Cannot Shuffle. Deck contains \(deck.count) cards")
            return
        }
        for _ in 0...1000 {
            let index = Int(arc4random_uniform(1000)) % deck.count
            let card = deck[index]
            deck.remove(at: index)
            deck.append(card)
        }
    }
    
    public func dealCard() -> Card? {
        if deck.count <= 0 {
            print("Cannot deal. Deck is empty!")
        }
        if let card = deck.first {
        deck.removeFirst()
         return card
        }
        return nil
    }
    
    public func size() -> Int {
        return deck.count
    }
}

File: Assignment 3/C# version/Exercise 9/Deck.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Card;

public class Deck
{
		private List<Card> deck;

		public Deck()
		{
			deck = new List<Card>();

			for (int suit = 0; suit < (int)Card.Suit.Count; suit++)
			{
				for (int rank = 0; rank < (int)Card.Rank.Count; rank++)
				{
					deck.Add(new Card((Card.Suit)suit, (Card.Rank)rank));
				}
			}
		}

		public void shuffle()
		{
			if (deck.Count != 52)
			{
				Console.Write("Cannot shuffle. Deck contains " + deck.Count + " cards.");
				return;
			}
			Random r = new Random();
			for (int i = 0; i < 1000; i++)
			{
				int ix = r.Next(deck.Count);
				Card c = deck.ElementAt(ix);
				deck.RemoveAt(ix);
				deck.Add(c);
			}
		}

		public int size()
		{
			return deck.Count;
		}

		public Card dealCard()
		{
			if (deck.Count <= 0)
			{
				Console.WriteLine("Cannot deal.Deck is empty!");
				return null;
			}
			Card c = deck.First();
			deck.RemoveAt(0);
			return c;
		}
	}